cmd/commandfuncs.go

```diff
@@ -538,6 +542,183 @@ func cmdValidateConfig(fl Flags) (int, error) {
 	return caddy.ExitCodeSuccess, nil
 }
 
+func cmdImportStorage(fl Flags) (int, error) {
+	importStorageCmdConfigFlag := fl.String("config")
+	importStorageCmdUserFlag := fl.String("user")
+  	importStorageCmdImportFile := fl.Arg(0)
+	
+	// re-run similarly to export here?
+	if importStorageCmdUserFlag != "" {
+	}
+
+  	config, _, err := LoadConfig(importStorageCmdConfigFlag, "")
+  	if err != nil {
+    		return caddy.ExitCodeFailedStartup, err
+  	}
+  	config = caddy.RemoveMetaFields(config)
+
+  	var cfg *caddy.Config
+  	err = caddy.StrictUnmarshalJSON(config, &cfg)
+  	if err != nil {
+    		return caddy.ExitCodeFailedStartup, err
+  	}
+
+	var input []byte
+
+  	// as a special case, read from stdin if the file name is "-"
+  	if importStorageCmdImportFile == "-" {
+		input, err = io.ReadAll(os.Stdin)
+		if err != nil {
+			return caddy.ExitCodeFailedStartup, fmt.Errorf("reading stdin: %v", err)
+		}
+	} else {
+		input, err = os.ReadFile(importStorageCmdImportFile)
+		if err != nil {
+			return caddy.ExitCodeFailedStartup, fmt.Errorf("reading input file: %v", err)
+		}
+	}
+
+  	err = caddy.ImportStorage(cfg, input)
+	if err != nil {
+		return caddy.ExitCodeFailedQuit, fmt.Errorf("importing storage: %v", err)
+	}
+	
+	fmt.Println("Successfully imported storage")
+  	return caddy.ExitCodeSuccess, nil
+}
+
+func cmdExportStorage(fl Flags) (int, error) {
+  	exportStorageCmdConfigFlag := fl.String("config")
+	exportStorageCmdOutputFlag := fl.String("output")
+	exportStorageCmdUserFlag := fl.String("user")
+	
+	var isFsStorage bool
+	standard, _, _, err := getModules()
+	if err != nil {
+		return caddy.ExitCodeFailedStartup, err
+	}	
+
+	for _, mod := range standard {
+		if mod.caddyModuleID == "caddy.storage.file_system" {
+			isFsStorage = true
+		}
+	}
+
+	u, err := user.Current()
+	if err != nil {
+		return caddy.ExitCodeFailedStartup, err 
+	}
+
+	// re-run command as the given user
+	// to ensure storage files are accessible.
+	// not sure this is necessary if sudo is needed
+	// anyways for setuid / setguid?
+	if exportStorageCmdUserFlag != "" && isFsStorage && u.Username != exportStorageCmdUserFlag {
+		u, err = user.Lookup(exportStorageCmdUserFlag)
+		if err != nil {
+			return caddy.ExitCodeFailedStartup, err 
+		}
+
+		uid, err := strconv.Atoi(u.Uid)
+		if err != nil {
+			return caddy.ExitCodeFailedStartup, err 
+		}
+		gid, err := strconv.Atoi(u.Gid) 
+		if err != nil {
+			return caddy.ExitCodeFailedStartup, err 
+		}
+
+		cmd := exec.Command("caddy", "storage-export", "--config", exportStorageCmdConfigFlag)
+		if exportStorageCmdOutputFlag != "" {
+			f, err := os.Create(exportStorageCmdOutputFlag)
+			if err != nil {
+				return caddy.ExitCodeFailedStartup, err
+			}
+			defer f.Close()
+		
+			cmd.Stdout = f
+		} else {
+			cmd.Stdout = os.Stdout
+		}
+		cmd.Stderr = os.Stderr
+		cmd.Env = []string{fmt.Sprint("HOME=", u.HomeDir)} 
```

Comment:
Rethinking this, I'm not sure we need `--user` at all. I think we should just add a mention that the command should be run as the same user Caddy runs as (typically `caddy` on Linux with systemd). Linux users can use `sudo su caddy -c 'caddy export -o <wherever>` to run it as the correct user.

On Windows, I think users typically end up running it as their current administrator user, so it shouldn't be a concern.

As you can see from the Windows CI build, it's failing because of the use of `syscall.Credential` which isn't defined on Windows. So I think we should avoid trying to muck with retrying as a different user.

Thanks for trying it though, it's an interesting experiment either way.

Reply:
Cool. I saw the CI failure and started looking at the Windows syscall interface (https://pkg.go.dev/golang.org/x/sys/windows#CreateProcessAsUser) but it's certainly simpler without all of that in the first place.